src/mux_client_impl/child.rs

```diff
@@ -0,0 +1,112 @@
+use super::{ChildStderr, ChildStdin, ChildStdout, Error};
+
+use core::mem::replace;
+
+use std::io;
+use std::os::unix::process::ExitStatusExt;
+use std::process::ExitStatus;
+
+use openssh_mux_client::connection::{EstablishedSession, SessionStatus};
+
+#[derive(Debug)]
+pub struct RemoteChild {
+    state: RemoteChildState,
+    child_stdin: Option<ChildStdin>,
+    child_stdout: Option<ChildStdout>,
+    child_stderr: Option<ChildStderr>,
+}
+
+impl RemoteChild {
+    pub(crate) fn new(
+        established_session: EstablishedSession,
+        child_stdin: Option<ChildStdin>,
+        child_stdout: Option<ChildStdout>,
+        child_stderr: Option<ChildStderr>,
+    ) -> Self {
+        Self {
+            state: RemoteChildState::Running(established_session),
+            child_stdin,
+            child_stdout,
+            child_stderr,
+        }
+    }
+
+    pub async fn disconnect(self) -> io::Result<()> {
+        use RemoteChildState::*;
+
+        match self.state {
+            Intermediate => unreachable!(),
+            Exited(_exit_status) => Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "invalid argument: can't kill an exited process",
+            )),
+            Running(_established_session) => Ok(()),
+        }
+    }
+
+    pub async fn wait(&mut self) -> Result<ExitStatus, Error> {
+        use RemoteChildState::*;
+
+        let exit_value = match self.state.take() {
+            Intermediate => unreachable!(),
+            Exited(exit_value) => exit_value,
+            Running(established_session) => match established_session.wait().await {
+                Ok(session_status) => match session_status {
+                    SessionStatus::TtyAllocFail(_established_session) => unreachable!(
+                        "openssh::mux_client_impl does not use feature tty by any means"
+                    ),
+                    SessionStatus::Exited { exit_value } => exit_value,
+                },
+                Err((err, established_session)) => {
+                    self.state = Running(established_session);
+                    return Err(err.into());
+                }
+            },
+        };
+
+        self.state = Exited(exit_value);
+
+        if let Some(val) = exit_value {
+            if val == 127 {
+                Err(Error::Remote(io::Error::new(
+                    io::ErrorKind::NotFound,
+                    "remote command not found",
+                )))
+            } else {
+                Ok(ExitStatusExt::from_raw((val as i32) << 8))
+            }
+        } else {
+            Err(Error::RemoteProcessTerminated)
+        }
+    }
+
+    pub fn try_wait(&mut self) -> Result<Option<ExitStatus>, Error> {
+        Ok(None)
```

Comment:
Hmm, it's going to be a bit tricky.

All the code in `openssh_mux_client` is `async` and they only return the whole response, implementing `try_wait` means I also have to add synchronous read and handle incomplete response.

Reply:
Yeah, that figures. I wonder if we should just remove `try_wait` entirely then :thinking:

Though I do also worry about what happens in `openssh_mux_client` then if a future is dropped before completing, such as with `select!` — does the client start seeing corrupted reads? Imagine something like:
```rust
tokio::select! {
  _ <- child.wait() => {},
  _ <- tokio::time::sleep(0) => {},
}
```
The `child` future will be waited on very briefly, and then dropped, and after that nothing stops a user from calling, say, `child.wait()` a second time.

Reply:
Luckily, the only function affected is the RemoteChild since it is stateful.
All other functions in Session and Command does not store any instance of openssh_mux_client::Connection, thus they are freed from this problem.

Even for RemoteChild, the solution is straightforward, because openssh_mux_client::EstablishedSession::wait takes “mut self” as value, instead of as reference, so if the future is cancelled, the entire EstablishedSession along with the Connection stored within it will be dropped, there will be no undefined state.

What I can do here is to add the documentation here regarding the cancel safety.

Reply:
I've updated doc on `RemoteChild::wait`.

I also checked mod `sftp` and it only uses `wait_with_output` which takes `self` by value, so it is totally OK.

Reply:
@jonhoo Shall I remove `try_wait` then?

On the other hand, I could implement such feature easily if `try_wait` can block or is `async`.

Since waiting for the remote child to exit is basically waiting for the socket to become readable, I can check the socket for readability.

If it is readable, then I would read the entire packet in blocking manner.
While this is indeed a blocking call, it is unlikely to block since `ssh` most likely would send it using one `write`/`send`, even if it does employ multiple `write`/`send`, these functions would just return immediately since the buffer for the unix socket is empty and should be big enough for one message.

If it is not readable, then I would return `Ok(None)`.

Does this sound good to you @jonhoo ?

Reply:
@jonhoo I have implemented `try_wait` based on the description above.